ctapipe/tools/stage1.py

```diff
@@ -0,0 +1,731 @@
+        product=meta.Product(
+            description="DL1 Data Product",
+            data_category="S",
+            data_level="DL1",
+            data_association="Subarray",
+            data_model_name="ASWG DL1",
+            data_model_version=DL1_DATA_MODEL_VERSION,
+            data_model_url="",
+            format="hdf5",
+        ),
+        process=meta.Process(type_="Simulation", subtype="", id_=int(obs_id)),
+        activity=meta.Activity.from_provenance(activity),
+        instrument=meta.Instrument(
+            site="Other",  # need a way to detect site...
+            class_="Subarray",
+            type_="unknown",
+            version="unknown",
+            id_=subarray.name,
+        ),
+    )
+
+    # convert all values to strings, since hdf5 can't handle Times, etc.:
+    # TODO: add activity_stop_time?
+    headers = {k: str(v) for k, v in reference.to_dict().items()}
+    meta.write_to_hdf5(headers, writer._h5file)
+
+
+def tel_type_string_to_int(tel_type):
+    """
+    convert a telescope type string (str(TelescopeDescription)) into an integer that
+    can be stored.
+
+    Parameters
+    ----------
+    tel_type: str
+        telescope type string like "SST_ASTRI_CHEC"
+
+    Returns
+    -------
+    int:
+        hash value
+    """
+    return np.int32(
+        int(hashlib.sha1(tel_type.encode("utf8")).hexdigest(), 16) % (10 ** 8)
+    )
+
+
+class ImageQualityQuery(QualityQuery):
+    """ for configuring image-wise data checks """
+
+    quality_criteria = List(
+        default_value=[
+            ("size_greater_0", "lambda image_selected: image_selected.sum() > 0"),
+        ],
+        help=QualityQuery.quality_criteria.help,
+    ).tag(config=True)
+
+
+class Stage1ProcessorTool(Tool):
+    name = "ctapipe-stage1-process"
+    description = __doc__ + f" This currently writes {DL1_DATA_MODEL_VERSION} DL1 data"
+    examples = """
+    To process data with all default values:
+    > ctapipe-stage1-process --input events.simtel.gz --output events.dl1.h5 --progress
+
+    Or use an external configuration file, where you can specify all options:
+    > ctapipe-stage1-process --config stage1_config.json --progress
+
+    The config file should be in JSON or python format (see traitlets docs). For an
+    example, see ctapipe/examples/stage1_config.json in the main code repo.
+    """
+
+    output_path = Path(
+        help="DL1 output filename", default_value=pathlib.Path("events.dl1.h5")
+    ).tag(config=True)
+
+    write_images = Bool(
+        help="Store DL1/Event/Image data in output", default_value=False
+    ).tag(config=True)
+
+    write_parameters = Bool(
+        help="Compute and store image parameters", default_value=True
+    ).tag(config=True)
+
+    compression_level = Int(
+        help="compression level, 0=None, 9=maximum", default_value=5, min=0, max=9
+    ).tag(config=True)
+
+    split_datasets_by = CaselessStrEnum(
```

Comment:
Do we want an option to not split the image parameter level?

Reply:
it was originally like that, but lots of people complained.  However, I'm happy to add it back. Then we definitely do need the tel_type field in the index container (there was a discussion to remove it)

Reply:
I don't see why we need that. This info can be joined in by telescope_id and via the subarray description.

Reply:
The point of the index for me is to enable such joining. It should be the minimal primary key. These files should be pretty normalized, shouldn't they?